data_safe_haven/config/config.py

```diff
@@ -1,199 +1,416 @@
-            blob_client.upload_blob(str(self), overwrite=True)
-            self.logger.info(
-                f"Uploaded config [green]{self.name}[/] to blob storage.",
+        return self.backend_
+
+    @property
+    def pulumi(self) -> ConfigSectionPulumi:
+        if not self.pulumi_:
+            self.pulumi_ = ConfigSectionPulumi()
+        return self.pulumi_
+
+    @property
+    def shm(self) -> ConfigSectionSHM:
+        if not self.shm_:
+            self.shm_ = ConfigSectionSHM(name=self.shm_name_)
+        return self.shm_
+
+    @property
+    def tags(self) -> ConfigSectionTags:
+        if not self.tags_:
+            self.tags_ = ConfigSectionTags(deployment=self.name)
+        return self.tags_
+
+    def __str__(self) -> str:
+        """String representation of the Config object"""
+        contents = {}
+        if self.azure_:
+            contents["azure"] = self.azure.to_dict()
+        if self.backend_:
+            contents["backend"] = self.backend.to_dict()
+        if self.pulumi_:
+            contents["pulumi"] = self.pulumi.to_dict()
+        if self.shm_:
+            contents["shm"] = self.shm.to_dict()
+        if self.sres:
+            contents["sre"] = {k: v.to_dict() for k, v in self.sres.items()}
+        if self.tags:
+            contents["tags"] = self.tags.to_dict()
+        return str(yaml.dump(contents, indent=2))
+
+    def read_stack(self, name: str, path: pathlib.Path):
+        """Add a Pulumi stack file to config"""
+        with open(path, "r", encoding="utf-8") as f_stack:
+            b64string = f_stack.read()
+        self.pulumi.stacks[name] = b64encode(b64string)
+
+    def remove_sre(self, name: str) -> None:
+        """Remove SRE config section by name"""
+        if name in self.sres.keys():
+            del self.sres[name]
+
+    def remove_stack(self, name: str) -> None:
+        """Remove Pulumi stack section by name"""
+        if name in self.pulumi.stacks.keys():
+            del self.pulumi.stacks[name]
+
+    def write_stack(self, name: str, path: pathlib.Path):
+        """Write a Pulumi stack file from config"""
+        b64string = b64decode(self.pulumi.stacks[name])
+        with open(path, "w", encoding="utf-8") as f_stack:
+            f_stack.write(b64string)
+
+    def download(self):
+        """Download config from Azure storage and interpret it"""
+        self.from_yaml(
+            yaml.safe_load(
+                self.azure_api.download_blob(
+                    self.filename,
+                    self.backend_resource_group_name,
+                    self.backend_storage_account_name,
+                    self.backend_storage_container_name,
+                )
             )
-        except Exception as exc:
-            raise DataSafeHavenAzureException(
-                f"Configuration file could not be uploaded to '{self.backend.storage_account_name}'."
-            ) from exc
+        )
+
+    def from_yaml(self, yaml_input: YamlType) -> None:
```

Comment:
Would it be better to have this as a classmethod?

Reply:
Maybe a better question is "should `from_yaml` and `download` just be moved into `__init__` since that's the only place they're used. What do you think?

Reply:
Done